cpp/include/cudf/detail/copy_range.cuh

```diff
@@ -0,0 +1,308 @@
+        atomicAdd(null_count, block_null_change);
+      }
+    }
+  }
+}
+
+}  // namespace anonymous
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+/**
+ * @brief Internal API to copy a range of values from a functor to a column.
+ *
+ * Copies N values from @p source to the range [@p begin, @p end)
+ * of @p target. @p target is modified in place.
+ *
+ * SourceFunctor must have these accessors:
+ * __device__ T data(cudf::size_type index);
+ * __device__ bool valid(cudf::size_type index);
+ *
+ * @tparam SourceFunctor the type of the source function object
+ * @param source An instance of SourceFunctor that provides data and valid mask
+ * @param target the column to copy into
+ * @param begin The beginning of the target range to write to
+ * @param end The index after the last element of the target range to write to
+ * @param stream CUDA stream to run this function
+ */
+template <typename SourceValueIterator, typename SourceValidityIterator>
+void copy_range(SourceValueIterator source_value_begin,
+                SourceValidityIterator source_validity_begin,
+                mutable_column_view& target,
+                size_type target_begin, size_type target_end,
+                cudaStream_t stream = 0) {
+  CUDF_EXPECTS((target_begin <= target_end) &&
+                 (target_begin >= 0) &&
+                 (target_begin < target.size()) &&
+                 (target_end <= target.size()),
+               "Range is out of bounds.");
+  using T = typename std::iterator_traits<SourceValueIterator>::value_type;
+
+  // this code assumes that source and target have the same type.
+  CUDF_EXPECTS(type_to_id<T>() == target.type().id(), "the data type mismatch");
+
+  static_assert(warp_size == cudf::util::size_in_bits<bitmask_type>(),
+    "copy_range_kernel assumes bitmask element size in bits == warp size");
+
+  T * __restrict__ data = target.head<T>();
+  bitmask_type * __restrict__ bitmask = target.null_mask();
+  auto offset = target.offset();
+
+#if 1
+  auto warp_aligned_begin_lower_bound =
+    size_type{target_begin - (target_begin % warp_size)};
+  auto warp_aligned_end_upper_bound =
+    size_type{
+      (target_end % warp_size) == 0 ? \
+        target_end : target_end + (warp_size - (target_end % warp_size))
+    };
+  auto num_items =
+    warp_aligned_end_upper_bound - warp_aligned_begin_lower_bound;
+#else
+  // This one results in a compiler internal error! TODO: file NVIDIA bug
+  // size_type num_items =
+  //   cudf::util::round_up_safe(target_end - target_begin, warp_size);
+  // number threads to cover range, rounded to nearest warp
+  // this code runs for one additional round if target_begin is not warp
+  // aligned, and target_end is block_size + 1
+  auto num_items =
+    size_type{
+      warp_size *
+        cudf::util::div_rounding_up_safe(target_end - target_begin, warp_size)
+    };
+#endif
+
+  constexpr auto block_size = int{256};
```

Comment:
This is coming from Herb Sutter's AAA (almost always auto) style (https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/), but if cudf discourages using this I will fix this.

I used this in multiple places, so I need to check the entire PR to eradicate this.

Reply:
Seems to be a link to a page about a podcast that doesn't have a link to the podcast?

I don't think we discourage this explicitly, but Jake's suggestion is definitely clearer and less redundant (and uses a cuDF type).

Reply:
OK, and I will eradicate _auto val = type{initial_value};_ style in my PR.

Reply:
And just FYI, I copy-and-pasted a block from the above AAA style link advocating "auto var = type{init_value};" style over "type var{init_value}", but I personally don't have a strong opinion about one style over the other (and I generally care much more about consistency in the codebase than specific coding styles), so I eradicated "auto var = type{init_value}" style from my PR.

> Second, commitment: “But in some cases I do want to commit to a specific type, not automatically deduce it, so I can’t use auto.” It’s true that sometimes you do want to commit to a specific type, but you can still use auto. As demonstrated in GotW #92 and #93, not only can you still write declarations of the form auto x = type{ init }; (instead of type x{init};) to commit to a specific type, but there are good reasons for doing so, such as that saying auto means you can’t possibly forget to initialize the variable.

> Guideline: Consider declaring local variables auto x = type{ expr }; when you do want to explicitly commit to a type. It is self-documenting to show that the code is explicitly requesting a conversion, it guarantees the variable will be initialized, and it won’t allow an accidental implicit narrowing conversion. Only when you do want explicit narrowing, use ( ) instead of { }.